File: prototype/OptimalControl/LowThrust/@LagrangeInterpolator/LagrangeInterpolator.m
classdef LagrangeInterpolator < handle
    
    %  Author: S. Hughes.  steven.p.hughes@nasa.gov
    %
    %  This class performs lagrange interpolation and differentiation using
    %  lagrange interpolating polynomials. 
    %
    %  Example Usage:  intepolate sin and exp functions and compute
    %  deriviatives at the points provided in the independent variables.
    %
    %      indVar  = [.1 .12 .13 .14 .15 .16 .17 .18]'
    %      interpValue = .125455;
    %      depVar1 = feval('sin',indVar);
    %      depVar2 = feval('exp',indVar);
    %      depVars = [depVar1 depVar2];
    %     
    %      lagInterp = LagrangeInterpolator();
    %      lagInterp.SetIndependentVariables(indVar);
    %      lagInterp.Interpolate(interpValue,depVars)
    %      lagInterp.Differentiate(depVars)
    
    properties  (SetAccess = 'protected')
        
        %   Column vector of independence variables.
        indVar
        %   The number of independent variables.  Order of interpolation
        %   will be 1-numPoints
        numPoints
        
    end
    
    methods
        
        %% Set the indepependent variables for the interpolator
        function SetIndependentVariables(this,indVar)
            
            %  Check that input is a column vector.
            [~,cols] = size(indVar);
            if  cols ~= 1
                errorMsg = 'Input must be a column vector';
                errorLoc  = ['LagrangeInterpolator:SetIndependent'...
                    'Variables'];
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            this.indVar    = indVar;
            this.numPoints = length(indVar);
            
        end
        
        %% Get the Lagrange products used in Lagrange Interpolation
        function lagProducts = GetLagrangeProducts(this,interpPoint)
            
            % Check that interpPoint is within range of indVar
            if interpPoint < this.indVar(1) || ...
                    interpPoint > this.indVar(this.numPoints)
                errorMsg = ['Requested interpolation point is not ' ...
                    'contained in vector of independent variables'];
                errorLoc = 'LagrangeInterpolator:GetLagrangeProducts';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            % Compute the Lagrange products
            lagProducts = ones(this.numPoints,1);
            for i = 1:this.numPoints
                for j = 1:this.numPoints
                    if i ~= j
                        lagProducts(i) = lagProducts(i)*(interpPoint - ...
                            this.indVar(j,1)) /(this.indVar(i,1) -...
                            this.indVar(j,1));
                    end
                end
            end
            
        end
        
        %%  Perform the interpolation
        function interpValues = Interpolate(this,interpPoint,funcValues)
            
            %  Check that the input is a column vector with correct number
            %  of rows
            [rows,cols] = size(funcValues);
            if rows ~= this.numPoints
                errorMsg = ['Number of rows in input must be the same' ...
                    'as the independent variable vector'];
                errorLoc = 'LagrangeInterpolator:Interpolate';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
                
            end
            
            %  Perform the interpolation
            interpValues = zeros(cols,1);
            lagProducts = GetLagrangeProducts(this,interpPoint);
            for funcIdx = 1:this.numPoints
                interpValues = interpValues + funcValues(funcIdx,:)'*...
                    lagProducts(funcIdx,1);
            end
            
        end
        
        %%  Compute the differentiation matrix
        function D = GetDiffMatrix(this)
            
            %  This code was taken from GPOPS v2.4
            n = this.numPoints;
            D = zeros(n,n);
            for j = 1:n;
                for i = 1:n;
                    prod = 1;
                    sum = 0;
                    if j == i
                        for k = 1:n
                            if k~=i
                                sum = sum+1/(this.indVar(i)...
                                    -this.indVar(k));
                            end
                        end
                        D(i,j) = sum;
                    else
                        for k = 1:n
                            if (k~=i) && (k~=j)
                                prod = prod*(this.indVar(i)- ...
                                    this.indVar(k));
                            end
                        end
                        for k = 1:n
                            if k~=j
                                prod = prod/(this.indVar(j)-...
                                    this.indVar(k));
                            end
                        end
                        D(i,j) = prod;
                    end
                end
            end
            
        end
        
        %  Compute derivatives of polynomial at the points provided in the
        %  independent variable vector
        function diffValues = Differentiate(this,funcValues)
            
            %  Check that the input is a column vector with correct number
            %  of rows
            [rows,~] = size(funcValues);
            if rows ~= this.numPoints
                errorMsg = ['Number of rows in input must be the same' ...
                    'as the independent variable vector'];
                errorLoc = 'LagrangeInterpolator:Differentiate';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            %   Perform the differentiation for each function 
            diffMat    = GetDiffMatrix(this);
            diffValues = diffMat*funcValues;

        end
        
    end
    
end

File: src/csaltTester/src/python/LagrangeInterpolator.py
import numpy as np

#Currently passing indVar into functions, rather than having a global variable.
#indVar = []

#def SetIndependentVariables(newIndVar):
#  indVar = newIndVar

order = 4      # Polynomial degree of the interpolation
startIndex = 0 # Index for the points used in the interpolation

class InterpolatorException(Exception):
   def __init__(self,*args,**kwargs):
      Exception.__init__(self,*args,**kwargs)

#Get the Lagrange products used in Lagrange Interpolation
def GetLagrangeProducts(interpPoint, indVar):
   global order
   global startIndex

   #Check that interpPoint is within range of indVar
   if interpPoint < indVar[0] or interpPoint > indVar[len(indVar) - 1]:
      raise InterpolatorException("Error: Requested interpolation point is not contained in the vector of independent variables")
   
   if order > len(indVar):
      order = len(indVar)

   #Compute the Lagrange products
   #lagProducts = np.ones(len(indVar))
   lagProducts = np.ones(order+1)
   
   # Set the start index so that the point is centered, as much as possible, in the range
   i = 0
   startIndex = 0
   while (indVar[i] < interpPoint):
      if i > order / 2 :
         startIndex = startIndex + 1
      i = i + 1

   if (startIndex + order + 1 > len(indVar)):
      startIndex = len(indVar) - order - 1

   # print("Start index for ", interpPoint, " = ", startIndex)

   i = 0
   while i <= order:  # len(indVar):
      j = 0
      while j <= order:  # len(indVar):
         if i != j:
            lagProducts[i] = lagProducts[i]*(interpPoint - indVar[startIndex + j])/(indVar[startIndex + i] - indVar[startIndex + j])
         j += 1
      i += 1

   return lagProducts

#Perform the interpolation
def Interpolate(interpPoint, funcValues, indVar):
   global order
   global startIndex

   rows = order + 1  # len(funcValues)
   cols = len(funcValues[0])

   #Check that the input is a column vector with correct number of rows
   # if rows != len(indVar):
   if rows > len(indVar):
      raise InterpolatorException("Error: Number of rows in input must not be more than the size of the independent variable vector")

   #Perform the interpolation
   interpValues = np.zeros(cols)
   # print("Interpolating to ", interpPoint)
   lagProducts = GetLagrangeProducts(interpPoint, indVar)

   # print("LagrangeProducts:  ", lagProducts)

   # funcIdx = 0
   funcIdx = startIndex
   lagIndex = 0
   while funcIdx < startIndex + order + 1:  # len(indVar):
      # print("   Function values: ", funcValues[funcIdx])
      interpValues = interpValues + np.array(funcValues[funcIdx])*lagProducts[lagIndex]
      funcIdx += 1
      lagIndex += 1

   # print("Interpolation values: ", interpValues)
   return interpValues

def InterpolateArray(interpPoints, funcValues, indVar):
   interpValues = np.zeros((len(interpPoints), len(funcValues[0])))

   rowIdx = 0
   while rowIdx < len(interpPoints):
      interpValues[rowIdx] = Interpolate(interpPoints[rowIdx], funcValues, indVar)
      rowIdx += 1

   interpValues = interpValues.tolist()
   return interpValues
